### DISCUSSION TITLE
Include custom fields in Meetings Export

### DISCUSSION THREAD

**Discussion by RockfaceIT** [code4recovery/12-step-meeting-list#1334] [Q&A] [answered]
Hello,

I have a few custom fields added via ACF and I would like them included in the meetings export.

Dashboard -> Meetings -> Import & Export -> Download CSV

Is there some code I can add to my functions.php?

Thank you in advance.

**Comment by kiyote33** [comment]
Hi @ckyoga, we don't have an easy way of doing what you are asking, as far as I know.  You could probably achieve what you are wanting to do by overwriting our tsml_get_meetings() function with your own custom code in the functions file. There you would need to retrieve your ACF data from the database for each $post and then appending that data to the $meeting array (sort of the same way the attendance_option is added).  Unfortunately the function code in question (tsml_get_meetings) is not wrapped with a function_exists() check so it can't be used in a child theme functions fille. Hope this helps...

We don't normally recommend doing this as your custom code would be overwritten when the plugin receives an update.

**Reply by RockfaceIT** [reply]
Darn.  Thank you for your prompt reply.  Do you know of any download/export plugins that would work with the Meetings datatype and ACF custom fields?  It seems pretty straight-forward.

Just to re-iterate -> I added a custom Post Type field for the 'Meetings' Post type called 'admin_comments'.  I want that field to be included in the export function & downloaded .csv file for each meeting download.

Thanks again!

**Reply by kiyote33** [reply]
No, I haven't had experience with anything meeting your specific requirements, but you might look at something related to exporting custom post types like 'WP All Export'.

**Comment by kiyote33** [comment]
Hi again @ckyoga, our team is looking at what may be a workable solution to your issue. We are running it through some tests and it may be available shortly if it works out.

**Comment by RockfaceIT** [comment]
That's Amazing!  Thank you...I look forward to your reply.

**Reply by kiyote33** [reply]
Hi @ckyoga, just a quick update here. We will be releasing 3.14.29 fairly soon with code changes to accommodate your request. We've tested it out and it seems workable. See 'enable exporting custom fields' PR #1335 to observe how it was setup prior to testing. Note that the $tsml_custom_meeting_fields array was placed in a child theme functions.php file to avoid being overridden when new releases are published. Hope this works for you. Let us know how it works out and if there's anything else you need.

**Comment by joshreisner** [comment]
hi @ckyoga yes once you get `3.14.29` if you add this to your theme's functions.php:

```
$tsml_custom_meeting_fields = [
    'admin_comments' => 'Admin Comments',
];
```

you should see that column in the CSV

**Reply by RockfaceIT** [reply]
Hello Josh,

I really appreciate you making this effort with us...so awesome!  I feel like we are very close!

I have updated the plugin and added the code to the child theme's functions.php. 
$tsml_custom_meeting_fields = [
    'admin_comments' => 'Admin Comments',
];

I can see the new column 'Admin Comments' in the export file but the actual field data is not coming thru. 

I edited two meetings with admin comments and neither meeting has the data in the export file.

please advise.

best,
Caroline Burns

[meetings-ACF-5.csv](https://github.com/code4recovery/12-step-meeting-list/files/14160255/meetings-ACF-5.csv)


<img width="647" alt="Screen Shot 2024-02-04 at 7 47 45 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/2f904959-db95-40ba-91a8-bff03755bb24">
<img width="958" alt="Screen Shot 2024-02-04 at 7 47 24 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/be6a7137-2a6e-476c-964c-408960c31781">

**Reply by joshreisner** [reply]
hi @ckyoga in your ACF field group settings, could you confirm the `name` of the Admin Comments field?

<img width="1727" alt="Screenshot 2024-02-04 at 9 27 22 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/1551689/fcd861a9-c487-43a5-b71b-a4e7df0bcab7">

**Reply by joshreisner** [reply]
oops i didn't see the other posts in the thread. it looks good! i'm not sure why you're not seeing the value show up. 

you saved changes on the meeting, and clicked the CSV again, and are not seeing the value?

<img width="376" alt="Screenshot 2024-02-04 at 9 34 52 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/1551689/ca32431e-7785-4c28-9c5d-ec8516e9e5c2">

**Comment by kiyote33** [comment]
Hi Caroline @ckyoga,
When editing the Field Group "admin_comment", did you assign it to specific meetings in the Settings/Rules like I did in the following graphic where you see my ACF Test field is assigned to the "Saturday Serenity" meeting?
![image](https://github.com/code4recovery/12-step-meeting-list/assets/57125427/b8261bed-9c14-4cf8-8abe-371a008d6fe9)
That will add whatever text is added in your Admin Comment field to only the Saturday Serenity meeting on export.

If you want the Admin Comments on every meeting, I suspect you would want your settings set like this:
![image](https://github.com/code4recovery/12-step-meeting-list/assets/57125427/f18a8197-76f5-4aa7-99a3-e547bf2fa95c)
You will still need to edit each meeting and add and save your admin comments before they will appear in the export file.

This appears to work correctly for us. Let us know @ckyoga if either of these suggestions work for you...

**Comment by RockfaceIT** [comment]
Hello,

We want the admin to have the ability to add admin_comments on any or all meetings and not just a specific meeting.

Here is our setup:

<img width="1148" alt="Screen Shot 2024-02-04 at 9 23 04 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/50743f5d-c60b-4be2-a428-81095071e4a0">
[meetings-ACF-5.csv](https://github.com/code4recovery/12-step-meeting-list/files/14160848/meetings-ACF-5.csv)

**Reply by kiyote33** [reply]
That  works! I had edited my message above to show that would be the correct setting to meet your needs. You may need  to refresh your screen.

**Comment by RockfaceIT** [comment]
Here is my field - is this right?

<img width="829" alt="Screen Shot 2024-02-04 at 9 29 53 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/8ec591f4-8d9f-4061-b454-389c7191c5c7">

**Reply by kiyote33** [reply]
Looks correct to me. Can you show me the Admin Comments of a meeting?

**Reply by kiyote33** [reply]
Hello Caroline. Are you still there?

**Comment by RockfaceIT** [comment]
yes, i am preparing a screenshot....one min

**Comment by RockfaceIT** [comment]
I am not seeing it.  Including CSV download, screenshot of plugin version, edit screen of test meeting
[meetings-ACF-6.csv](https://github.com/code4recovery/12-step-meeting-list/files/14160932/meetings-ACF-6.csv)

<img width="704" alt="Screen Shot 2024-02-04 at 9 37 28 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/1afd984b-fd88-4b75-aad0-a63647cd8282">
<img width="882" alt="Screen Shot 2024-02-04 at 9 36 46 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/afcdd7d9-edbd-4ba5-980b-01b21580e570">

**Reply by kiyote33** [reply]
If you entered the "test admin comment....in new meeting" text and then published the meeting, you should see it in the export meetings.csv file for the Test ACF field meeting. You're confirming that did not happen?

**Reply by kiyote33** [reply]
What does the line of code entered in your theme functions.php file look like?

**Reply by RockfaceIT** [reply]
<img width="722" alt="Screen Shot 2024-02-04 at 9 47 05 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/05bf146d-b558-41a1-897f-ac7cb2b27d1c">
<img width="1133" alt="Screen Shot 2024-02-04 at 9 44 58 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/2d5effa7-d439-4955-8e8b-3a385fb56809">

**Reply by kiyote33** [reply]
It's perfect. Give me a minute to review. I must have missed something because everything on first pass seems correct.

**Comment by RockfaceIT** [comment]
You are right...the admin_comments did not export.

**Reply by kiyote33** [reply]
Can you try editing the meeting, add "Hello World" to the beginning of the Admin Comments, hit the Update button and export the csv again. I'm wondering if maybe the text isn't saving for some  reason.

**Reply by RockfaceIT** [reply]
I added 'hello world' to the beginning and still no change.  The 'Admin Comments' header is in the file but the rest of the column is blank, the data does not get exported.  Including the button I am using for download and the ACF field details.

<img width="1086" alt="Screen Shot 2024-02-04 at 10 00 36 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/ef05aeaf-4568-438f-b7a7-9c5071c4d69b">
  
<img width="621" alt="Screen Shot 2024-02-04 at 10 00 27 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/c1806a32-ee92-49d1-a378-23d4a3650424">

[meetings.csv](https://github.com/code4recovery/12-step-meeting-list/files/14161100/meetings.csv)


<img width="582" alt="Screen Shot 2024-02-04 at 10 00 50 PM" src="https://github.com/code4recovery/12-step-meeting-list/assets/37557944/52898ed1-91e4-49c2-909e-2d3de24b1064">

**Reply by kiyote33** [reply]
Everything you've done looks absolutely correct. I've think I have enough information to duplicate exactly what you are doing. Give me 15-20 minutes to duplicate your process and maybe I can come up with something.

**Reply by RockfaceIT** [reply]
sounds good.  let me know if you want to take a look at our wp-admin

**Reply by kiyote33** [reply]
I'll do that next if this doesn't work out.

**Reply by kiyote33** [reply]
![image](https://github.com/code4recovery/12-step-meeting-list/assets/57125427/2c816b04-342f-4cc1-b092-46d8c434128e)

Ok. I replicated your process and it worked perfect. I can confirm it's not anything wrong with your process. Let's take a look at your site if you'd like. Can you give me some temporary credentials?

**Reply by kiyote33** [reply]
you can send them to me at waynep@code4recovery.org

**Reply by kiyote33** [reply]
It's midnight here, so I'm signing off for tonight. We can continue when it's convenient for you tomorrow. Good nite...

**Reply by RockfaceIT** [reply]
Sent you an email from 'ckyoga@gmail.com'

**Reply by kiyote33** [reply]
Hi @ckyoga, Thanks, I got your email. I'll take a look-see and let you know what turns up, if anything.

**Comment by kiyote33** [comment]
Caroline, I added the Hello theme to my test site along with a child Elementor  theme for it.  The admin_comments are still working correctly.  So I think the theme you are using is OK. I do see there is an update available for Elementor Pro. Can we update that or has the license expired?

**Reply by RockfaceIT** [reply]
interesting....elementor pro is updated now.  You are seeing the actual data in the column?

**Comment by kiyote33** [comment]
Life is sweet again, eh!

**Comment by kiyote33** [comment]
Ah! you were asking a question. No, I see no joy.

**Reply by RockfaceIT** [reply]
oh darn!

**Comment by kiyote33** [comment]
Hi Caroline, My test site pretty much mimics your site, except I'm getting the expected results in the meetings.csv downloads.  I can see no obvious reason why your site isn't getting the same. I looked in the json cache file and can see your data for the admin column is not there.

I'm sort of grasping at straws here. If you don't mind, I would like to restart from the beginning with a clean state. Sort of like a reboot where I recreate the test group and field. There has to be something I'm missing.

**Reply by RockfaceIT** [reply]
Geez....have you tried uploading my meetings.csv into your test setup?  Maybe the way the data is laid out affects the export.  I am not sure.

I will also try on a fresh install.  This site is pretty simple as of yet.  Let me know how I can assist.

**Reply by kiyote33** [reply]
Unfortunately the problem didn't transfer to my test machine when I uploaded your data.

**Reply by RockfaceIT** [reply]
shoot....Thank you for trying.  I am going to have to start fresh as well to see if I can figure this out....very odd.  I will get back to you later today.

**Reply by kiyote33** [reply, marked as answer]
So between Josh & me we sort of figured out that the two piece label when setting up ACF in your environment was not working. We added 'test' => 'Test' which worked fine, but after many fruitless attempts to get 'admin_comments' => Admin Comments' to work we settled on 'comments' => 'Comments' and that worked. There seems to be something in your environment that is causing our code to bomb when it sees the two-word label, or a hyphen or underscore in the ACF field name.  Weird. But I think you've now got a work-around. Let us know if that will work for you...

**Reply by joshreisner** [reply]
yeah i wonder if there is some database cruft on the admin_comments name specifically, if it would work for example if it were admin_notes? if you learn anything please let us know!

**Reply by RockfaceIT** [reply]
Phew....good find! 

I tried the following in functions.php with the ACF field definition matching.

**Did not work:**
$tsml_custom_meeting_fields = [ 'test' => 'Test', 'comments' => 'Comments', 'admin_test' => 'Admin Test'];
$tsml_custom_meeting_fields = [ 'test' => 'Test', 'comments' => 'Comments', 'admin_test' => 'AdminTest'];
$tsml_custom_meeting_fields = [ 'test' => 'Test', 'comments' => 'Comments', 'admintest' => 'Admin Test'];

**Worked** - Must remove all spaces and special characters to get it to work.  This is great!
$tsml_custom_meeting_fields = [ 'test' => 'Test', 'comments' => 'Comments', 'admintest' => 'AdminTest'];

Once again, thank you sooo soo much!

**Reply by joshreisner** [reply]
oops i see the problem! this is probably TMI but [in this function we are array_mapping](https://github.com/code4recovery/12-step-meeting-list/blob/main/includes/functions.php#L1345) on a multidimensional array - we should be using array_keys.

this will be fixed in [this PR](https://github.com/code4recovery/12-step-meeting-list/pull/1336), which is probably why i wasn't able to recreate the issue -- i think i was probably testing on a site running that branch 😅

**Reply by RockfaceIT** [reply]
Ha!   Not TMI...I know PHP.  Thanks again!

**Comment by bubfcg** [comment]
Just came across this topic. I also added custom fields with ACF. I do get them in my export. But what I want to do is fill the custom fields with the CSV. But That doesn't seem to work. 

1. I made an export, the custom fields were in it
2. I deleted all the meetings in the export
3. Added 1 new test meeting, with values in the custom fields
4. Uploaded the csv, the new meeting was added, but custom fields were empty

![image](https://github.com/code4recovery/12-step-meeting-list/assets/167124715/0121d0cb-b507-4568-b663-a1af29a8319d)
